lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,332 @@
+          var redirectUri = this.buildSuccessRedirectUri(uri, responseType);
+
+          this.updateResponse(response, redirectUri, state);
+
+          return code;
+        })
+        .catch(function(e) {
+          if (!(e instanceof OAuthError)) {
+            e = new ServerError(e);
+          }
+
+          var redirectUri = this.buildErrorRedirectUri(uri, e);
+
+          this.updateResponse(response, redirectUri, state);
+
+          throw e;
+        });
+    });
+};
+
+/**
+ * Generate authorization code.
+ */
+
+AuthorizeHandler.prototype.generateAuthorizationCode = function() {
+  if (this.model.generateAuthorizationCode) {
+    return Promise.try(this.model.generateAuthorizationCode);
+  }
+
+  return tokenUtil.generateRandomToken();
+};
+
+/**
+ * Get authorization code lifetime.
+ */
+
+AuthorizeHandler.prototype.getAuthorizationCodeLifetime = function() {
+  var expires = new Date();
+
+  expires.setSeconds(expires.getSeconds() + this.authorizationCodeLifetime);
+
+  return expires;
+};
+
+/**
+ * Get the client from the model.
+ */
+
+AuthorizeHandler.prototype.getClient = function(request) {
+  var clientId = request.body.client_id || request.query.client_id;
+
+  if (!clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!is.vschar(clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  var redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+  if (redirectUri && !is.uri(redirectUri)) {
+    throw new InvalidRequestError('Invalid request: `redirect_uri` is not a valid URI');
+  }
+
+  return Promise.try(this.model.getClient, clientId)
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client credentials are invalid');
+      }
+
+      if (!client.grants) {
+        throw new InvalidClientError('Invalid client: missing client `grants`');
+      }
+
+      if (!_.contains(client.grants, 'authorization_code')) {
+        throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+      }
+
+      if (!client.redirectUris || 0 === client.redirectUris.length) {
+        throw new InvalidClientError('Invalid client: missing client `redirectUri`');
+      }
+
+      if (redirectUri && !_.contains(client.redirectUris, redirectUri)) {
+        throw new InvalidClientError('Invalid client: `redirect_uri` does not match client value');
+      }
+
+      return client;
+    });
+};
+
+/**
+ * Get scope from the request.
+ */
+
+AuthorizeHandler.prototype.getScope = function(request) {
+  var scope = request.body.scope || request.query.scope;
+
+  if (!is.nqschar(scope)) {
+    throw new InvalidScopeError('Invalid parameter: `scope`');
+  }
+
+  return scope;
+};
+
+/**
+ * Get state from the request.
+ */
+
+AuthorizeHandler.prototype.getState = function(request) {
+  var state = request.body.state || request.query.state;
+
+  if (!state) {
```

Comment:
State is indeed (and unfortunately) an optional parameter. It is my contention (and many others') that `state` is an absolutely necessary feature for a proper OAuth2 server implementation (it is also recommended by the [spec](https://tools.ietf.org/html/rfc6749#section-4.1.1)).

In 99% of the cases, the `state` parameter is the only thing protecting you from a [CSRF](https://tools.ietf.org/html/rfc6749#section-10.12) attack. The other %1, well, let's just say I have yet to encounter it.

I believe it is the responsibility of this package to educate their users on proper security and requiring `state` is essential.

Note that a lot more could be done here: I could be requiring `state` to be a random, non-guessable, unique value (as I recommend it should be), but in fact, anyone can simply pass `state=foobar` and get away with it.

It is my hope that making it _required_ will force our users to read up on the subject and properly secure their applications.